ee/agent/storage/sqlite/keyvalue_store_sqlite.go

```diff
@@ -0,0 +1,300 @@
+	conn, err := sql.Open("sqlite", startupDbFilepath)
+	if err != nil {
+		return nil, fmt.Errorf("opening startup db in %s: %w", rootDirectory, err)
+	}
+	if err := conn.PingContext(ctx); err != nil {
+		return nil, fmt.Errorf("establishing valid connection to startup db: ping error: %w", err)
+	}
+
+	return conn, nil
+}
+
+// validateDb ensures that the database file exists and is not corrupt.
+func validateDb(ctx context.Context, dbFilepath string) error {
+	// Make sure the database exists
+	if _, err := os.Stat(dbFilepath); os.IsNotExist(err) {
+		return fmt.Errorf("db does not exist at %s", dbFilepath)
+	} else if err != nil {
+		return fmt.Errorf("determining if db exists: %w", err)
+	}
+
+	conn, err := sql.Open("sqlite", dbFilepath)
+	if err != nil {
+		return fmt.Errorf("creating connection: %w", err)
+	}
+	defer conn.Close()
+
+	// Make sure the database is valid
+	var quickCheckResults string
+	if err := conn.QueryRowContext(ctx, `pragma quick_check;`).Scan(&quickCheckResults); err != nil {
+		return fmt.Errorf("running quick check: %w", err)
+	}
+	if quickCheckResults != "ok" {
+		return fmt.Errorf("quick check did not pass: %s", quickCheckResults)
+	}
+
+	return nil
+}
+
+// dbLocation standardizes the filepath to the given database.
+func dbLocation(rootDirectory string) string {
+	// Note that the migration framework expects a net/url style path,
+	// so we adjust the rootDirectory with filepath.ToSlash and then
+	// use path.Join instead of filepath.Join here.
+	return path.Join(filepath.ToSlash(rootDirectory), "kv.sqlite")
+}
+
+// migrate makes sure that the database schema is correct.
+func (s *sqliteStore) migrate(ctx context.Context) error {
+	d, err := iofs.New(migrations, "migrations")
+	if err != nil {
+		return fmt.Errorf("loading migration files: %w", err)
+	}
+	defer d.Close()
+
+	dbInstance, err := sqlitemigrationdriver.WithInstance(s.conn, &sqlitemigrationdriver.Config{})
+	if err != nil {
+		return fmt.Errorf("creating db migration instance: %w", err)
+	}
+	m, err := migrate.NewWithInstance("iofs", d, "sqlite", dbInstance)
+	if err != nil {
+		return fmt.Errorf("creating migrate instance: %w", err)
+	}
+
+	if err := m.Up(); err != nil && !errors.Is(err, migrate.ErrNoChange) {
+		return fmt.Errorf("running migrations: %w", err)
+	}
+
+	return nil
+}
+
+func (s *sqliteStore) Close() error {
+	return s.conn.Close()
+}
+
+func (s *sqliteStore) Get(key []byte) (value []byte, err error) {
+	if s == nil {
+		return nil, errors.New("store is nil")
+	}
+
+	// It's fine to interpolate the table name into the query because
+	// we require the table name to be in our allowlist `supportedTables`
+	query := fmt.Sprintf(`SELECT value FROM %s WHERE name = ?;`, s.tableName)
+
+	var keyValue string
+	if err := s.conn.QueryRow(query, string(key)).Scan(&keyValue); err != nil {
+		return nil, fmt.Errorf("querying key `%s`: %w", string(key), err)
+	}
+	return []byte(keyValue), nil
+}
+
+func (s *sqliteStore) Set(key, value []byte) error {
```

Comment:
Consider something like `s.readOnly` as a check around `OpenRO` creating the DB and allowing writes.